src/Stratis.Bitcoin/Consensus/PHConsensusManagerBehavior.cs

```diff
@@ -0,0 +1,306 @@
+
+                    this.logger.LogTrace("{0} headers were added to cache, new cache size is {1}.", headers.Count, this.cachedHeaders.Count);
+                    this.logger.LogTrace("(-)[CACHED]");
+                    return;
+                }
+
+                var consensusFactory = new PosConsensusFactory();
+
+                ConnectNewHeadersResult result = await this.PresentHeadersLockedAsync(headers).ConfigureAwait(false);
+
+                if (result == null)
+                {
+                    this.logger.LogTrace("Processing of {0} headers failed.", headers.Count);
+                    this.logger.LogTrace("(-)[PROCESSING_FAILED]");
+                    return;
+                }
+
+                this.ExpectedPeerTip = result.Consumed;
+                this.UpdateBestSentHeader(this.ExpectedPeerTip);
+
+                if (result.Consumed.HashBlock != headers.Last().GetHash())
+                {
+                    var posBock = new PosBlock(result.Consumed.Header);
+                    ProvenBlockHeader provenBlockHeader = consensusFactory.CreateProvenBlockHeader(posBock);
+                    int consumedCount = headers.IndexOf(provenBlockHeader) + 1;
+                    this.cachedHeaders.AddRange(headers.Skip(consumedCount));
+
+                    this.logger.LogTrace("{0} out of {1} items were not consumed and added to cache.", headers.Count - consumedCount, headers.Count);
+                }
+
+                if (this.cachedHeaders.Count == 0)
+                    await this.ResyncAsync().ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>Constructs the proven headers payload from locator to consensus tip.</summary>
+        /// <param name="locator">Block locator.</param>
+        /// <param name="hashStop">Hash of the block after which constructing headers payload should stop.</param>
+        /// <param name="lastHeader"><see cref="ProvenBlockHeader"/> of the last header that was added to the <see cref="ProvenHeadersPayload"/>.</param>
+        /// <returns><see cref="ProvenHeadersPayload"/> with headers from locator towards consensus tip or <c>null</c> in case locator was invalid.</returns>
+        private ProvenHeadersPayload ConstructHeadersPayload(BlockLocator locator, uint256 hashStop, out ProvenBlockHeader lastHeader)
+        {
+            ChainedHeader fork = this.chain.FindFork(locator);
+
+            lastHeader = null;
+
+            if (fork == null)
+            {
+                this.logger.LogTrace("(-)[INVALID_LOCATOR]:null");
+                return null;
+            }
+
+            var headers = new ProvenHeadersPayload();
+            var consensusFactory = new PosConsensusFactory();
+            foreach (ChainedHeader header in this.chain.EnumerateToTip(fork).Skip(1))
+            {
+                var posBock = new PosBlock(header.Header);
+                ProvenBlockHeader provenBlockHeader = consensusFactory.CreateProvenBlockHeader(posBock);
```

Comment:
don't create it, load it from PH store
Load all PH headers from store in 1 batch